node/src/components/block_synchronizer/tests.rs

```diff
@@ -336,3 +452,1231 @@ async fn should_not_stall_after_registering_new_era_validator_weights() {
+                                                                * signatures */
+    )
+    .await;
+
+    // Check what signatures were requested
+    let mut sigs_requested = Vec::new();
+    for event in events {
+        assert_matches!(
+            event,
+            MockReactorEvent::FinalitySignatureFetcherRequest(FetcherRequest {
+                id,
+                peer,
+                ..
+            }) => {
+                assert!(peers.contains(&peer));
+                assert_eq!(id.block_hash, *block.hash());
+                assert_eq!(id.era_id, block.header().era_id());
+                sigs_requested.push((peer, id.public_key));
+            }
+        );
+    }
+
+    // Simulate failed fetch of finality signatures
+    let mut generated_effects = Effects::new();
+    for (peer, public_key) in sigs_requested {
+        let effects = block_synchronizer.handle_event(
+            mock_reactor.effect_builder(),
+            &mut rng,
+            Event::FinalitySignatureFetched(Err(FetcherError::Absent {
+                id: Box::new(Box::new(FinalitySignatureId {
+                    block_hash: *block.hash(),
+                    era_id: block.header().era_id(),
+                    public_key,
+                })),
+                peer,
+            })),
+        );
+        // the effects array should be empty while the latch is active
+        // once the latch is reset, we should get some effects
+        generated_effects.extend(effects);
+    }
+
+    // If the effects are empty at this point, then the synchronizer gets stuck
+    assert!(!generated_effects.is_empty());
+}
+
+#[tokio::test]
+async fn next_action_for_have_weak_finality_is_fetching_block_body() {
+    let mut rng = TestRng::new();
+    let mock_reactor = MockReactor::new();
+    let test_env = TestEnv::random(&mut rng);
+    let peers = test_env.peers();
+    let block = test_env.block();
+    let validator_matrix = test_env.gen_validator_matrix();
+    let validators_private_keys = test_env.validator_keys();
+    let mut block_synchronizer = BlockSynchronizer::new_initialized(&mut rng, validator_matrix);
+
+    // Register block for fwd sync
+    assert!(block_synchronizer.register_block_by_hash(*block.hash(), false, true));
+    assert!(block_synchronizer.forward.is_some());
+    block_synchronizer.register_peers(*block.hash(), peers.clone());
+
+    let fwd_builder = block_synchronizer
+        .forward
+        .as_mut()
+        .expect("Forward builder should have been initialized");
+    assert!(fwd_builder
+        .register_block_header(block.clone().take_header(), None)
+        .is_ok());
+    fwd_builder.register_era_validator_weights(&block_synchronizer.validator_matrix);
+
+    // Register finality signatures to reach weak finality
+    register_multiple_signatures(
+        fwd_builder,
+        block,
+        validators_private_keys
+            .iter()
+            .take(validators_private_keys.len() / 3 + 1),
```

Comment:
Can you use `let num_sigs_required_for_weak_finality = validators_private_keys.len() / 3` here, as in other tests?

Reply:
Actually adding one extra sig is correct to account for the lossy integer division. I'll update the test that didn't have this criteria.